File: src/lexer/lexer.ts
export type TokenType = "IMPORT" | "FROM" | "INTERFACE" | "COMPONENT" | "EXTENDS" | "IMPLEMENTS" | "TYPE" | "IDENTIFIER" | "STRING" | "NUMBER" | "BOOLEAN" | "NULL" | "COLON" | "QUESTION" | "EQUALS" | "LPAREN" | "RPAREN" | "LBRACE" | "RBRACE" | "PIPE" | "COMMA" | "INDENT" | "DEDENT" | "NEWLINE" | "EOF";
export interface Token { type: TokenType; value: string; line: number; column: number; }
export class TamlLexError extends Error { constructor(public readonly line: number, public readonly column: number, message: string) { super(`[TAML Lex Error] ${line}:${column}: ${message}`); } }

const keywords: Record<string, TokenType> = { import: "IMPORT", from: "FROM", interface: "INTERFACE", component: "COMPONENT", extends: "EXTENDS", implements: "IMPLEMENTS" };
const types = new Set(["String", "Int", "Float", "Boolean"]);
export class TamlLexer {
  constructor(private readonly source: string) {}
  tokenize(): Token[] {
    const tokens: Token[] = []; const indents = [0];
    const lines = this.source.replace(/^\uFEFF/, "").replace(/\r\n?/g, "\n").split("\n");
    for (let index = 0; index < lines.length; index++) {
      const line = index + 1; const raw = lines[index]; let position = 0;
      while (raw[position] === " ") position++;
      if (raw[position] === "\t") throw new TamlLexError(line, position + 1, "tabs are not allowed; use spaces");
      const content = raw.slice(position);
      if (content === "" || content.startsWith("#")) continue;
      if (position > indents[indents.length - 1]) { indents.push(position); tokens.push({ type: "INDENT", value: "", line, column: 1 }); }
      else { while (position < indents[indents.length - 1]) { indents.pop(); tokens.push({ type: "DEDENT", value: "", line, column: 1 }); } if (position !== indents[indents.length - 1]) throw new TamlLexError(line, 1, "inconsistent indentation"); }
      while (position < raw.length) {
        const char = raw[position]; const col = position + 1;
        if (char === " " || char === "\t") { position++; continue; }
        if (char === "#") break;
        const punctuator: Record<string, TokenType> = { ":": "COLON", "?": "QUESTION", "=": "EQUALS", "(": "LPAREN", ")": "RPAREN", "{": "LBRACE", "}": "RBRACE", "|": "PIPE", ",": "COMMA" };
        if (punctuator[char]) { tokens.push({ type: punctuator[char], value: char, line, column: col }); position++; continue; }
        if (char === '"' || char === "'") { const quote = char; let value = ""; position++; let closed = false; while (position < raw.length) { const c = raw[position++]; if (c === quote) { closed = true; break; } if (c === "\\" && quote === '"') { const escaped = raw[position++]; const map: Record<string, string> = { n: "\n", r: "\r", t: "\t", '"': '"', "\\": "\\" }; if (escaped === undefined || map[escaped] === undefined) throw new TamlLexError(line, position, "invalid string escape"); value += map[escaped]; } else value += c; } if (!closed) throw new TamlLexError(line, col, "unterminated string literal"); tokens.push({ type: "STRING", value, line, column: col }); continue; }
        const number = /^-?(?:0|[1-9]\d*)(?:\.\d+)?/.exec(raw.slice(position));
        if (number) { tokens.push({ type: "NUMBER", value: number[0], line, column: col }); position += number[0].length; continue; }
        const word = /^[A-Za-z_][A-Za-z0-9_-]*/.exec(raw.slice(position));
        if (word) { const value = word[0]; const type: TokenType = keywords[value] ?? (types.has(value) ? "TYPE" : value === "true" || value === "false" ? "BOOLEAN" : value === "null" ? "NULL" : "IDENTIFIER"); tokens.push({ type, value, line, column: col }); position += value.length; continue; }
        throw new TamlLexError(line, col, `invalid character '${char}'`);
      }
      tokens.push({ type: "NEWLINE", value: "", line, column: raw.length + 1 });
    }
    while (indents.length > 1) { indents.pop(); tokens.push({ type: "DEDENT", value: "", line: lines.length, column: 1 }); }
    tokens.push({ type: "EOF", value: "", line: lines.length, column: 1 }); return tokens;
  }
}

File: crates/taml-core/src/lexer.rs
use crate::TamlDiagnostic;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum TokenKind {
    Interface,
    Component,
    Extends,
    Implements,
    Identifier,
    String,
    Number,
    Boolean,
    Colon,
    Question,
    Equals,
    LParen,
    RParen,
    Pipe,
    Comma,
    Newline,
    Indent,
    Dedent,
    Eof,
}
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct Token<'a> {
    pub kind: TokenKind,
    pub text: &'a str,
    pub offset: usize,
    pub line: usize,
    pub column: usize,
}

/// Zero-copy scanner: every token borrows directly from the source buffer.
pub fn lex(source: &str) -> Result<Vec<Token<'_>>, TamlDiagnostic> {
    let mut output = Vec::new();
    let mut indents = vec![0usize];
    let mut offset = 0usize;
    for (line_index, line) in source.lines().enumerate() {
        let line_number = line_index + 1;
        let content = line.split('#').next().unwrap_or("");
        let spaces = content.len() - content.trim_start_matches(' ').len();
        if content.trim().is_empty() {
            offset += line.len() + 1;
            continue;
        }
        if content.as_bytes().get(spaces) == Some(&b'\t') {
            return Err(TamlDiagnostic::at(
                "LEX_ERROR",
                "tabs are not allowed",
                offset + spaces,
                line_number,
                spaces + 1,
            ));
        }
        if spaces > *indents.last().unwrap_or(&0) {
            indents.push(spaces);
            output.push(Token {
                kind: TokenKind::Indent,
                text: "",
                offset,
                line: line_number,
                column: 1,
            });
        }
        while spaces < *indents.last().unwrap_or(&0) {
            indents.pop();
            output.push(Token {
                kind: TokenKind::Dedent,
                text: "",
                offset,
                line: line_number,
                column: 1,
            });
        }
        if spaces != *indents.last().unwrap_or(&0) {
            return Err(TamlDiagnostic::at(
                "LEX_ERROR",
                "inconsistent indentation",
                offset,
                line_number,
                1,
            ));
        }
        let bytes = content.as_bytes();
        let mut cursor = spaces;
        while cursor < bytes.len() {
            if bytes[cursor].is_ascii_whitespace() {
                cursor += 1;
                continue;
            }
            let start = cursor;
            let kind = match bytes[cursor] {
                b':' => {
                    cursor += 1;
                    TokenKind::Colon
                }
                b'?' => {
                    cursor += 1;
                    TokenKind::Question
                }
                b'=' => {
                    cursor += 1;
                    TokenKind::Equals
                }
                b'(' => {
                    cursor += 1;
                    TokenKind::LParen
                }
                b')' => {
                    cursor += 1;
                    TokenKind::RParen
                }
                b'|' => {
                    cursor += 1;
                    TokenKind::Pipe
                }
                b',' => {
                    cursor += 1;
                    TokenKind::Comma
                }
                b'"' | b'\'' => {
                    let quote = bytes[cursor];
                    cursor += 1;
                    while cursor < bytes.len() && bytes[cursor] != quote {
                        if bytes[cursor] == b'\\' && quote == b'"' {
                            cursor += 1;
                        }
                        cursor += 1;
                    }
                    if cursor == bytes.len() {
                        return Err(TamlDiagnostic::at(
                            "LEX_ERROR",
                            "unterminated string",
                            offset + start,
                            line_number,
                            start + 1,
                        ));
                    }
                    cursor += 1;
                    TokenKind::String
                }
                b'-' | b'0'..=b'9' => {
                    cursor += 1;
                    while cursor < bytes.len()
                        && (bytes[cursor].is_ascii_digit() || bytes[cursor] == b'.')
                    {
                        cursor += 1;
                    }
                    TokenKind::Number
                }
                b'A'..=b'Z' | b'a'..=b'z' | b'_' => {
                    cursor += 1;
                    while cursor < bytes.len()
                        && (bytes[cursor].is_ascii_alphanumeric()
                            || matches!(bytes[cursor], b'_' | b'-'))
                    {
                        cursor += 1;
                    }
                    match &content[start..cursor] {
                        "interface" => TokenKind::Interface,
                        "component" => TokenKind::Component,
                        "extends" => TokenKind::Extends,
                        "implements" => TokenKind::Implements,
                        "true" | "false" => TokenKind::Boolean,
                        _ => TokenKind::Identifier,
                    }
                }
                _ => {
                    return Err(TamlDiagnostic::at(
                        "LEX_ERROR",
                        "invalid character",
                        offset + cursor,
                        line_number,
                        cursor + 1,
                    ));
                }
            };
            output.push(Token {
                kind,
                text: &content[start..cursor],
                offset: offset + start,
                line: line_number,
                column: start + 1,
            });
        }
        output.push(Token {
            kind: TokenKind::Newline,
            text: "",
            offset: offset + line.len(),
            line: line_number,
            column: line.len() + 1,
        });
        offset += line.len() + 1;
    }
    while indents.len() > 1 {
        indents.pop();
        output.push(Token {
            kind: TokenKind::Dedent,
            text: "",
            offset,
            line: source.lines().count() + 1,
            column: 1,
        });
    }
    output.push(Token {
        kind: TokenKind::Eof,
        text: "",
        offset: source.len(),
        line: source.lines().count() + 1,
        column: 1,
    });
    Ok(output)
}
